From 5246f5e3d267543f5778ed233e6d601757e3b2ed Mon Sep 17 00:00:00 2001
From: Prabal Singh <prabalsingh@google.com>
Date: Fri, 8 May 2026 09:19:52 +0000
Subject: [PATCH 28/49] Set min_sdk_version in GooglePackageInstaller

This change sets the min_sdk_version for GooglePackageInstaller to 36. This also adds some checks to make sure APIs are only called when they are available.

Test: manually tested
Bug: 507488454
Flag: EXEMPT BUGFIX
(cherry picked from commit bbd5d104d3024e33023d2a7495b189c6b647235f)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:96792b5de9a2e1d2958e3cfaf61d0cb274fd48e8
Merged-In: I5d6640efb0713e4ee1c2f1eecff1a1df4cc6c858
Change-Id: I5d6640efb0713e4ee1c2f1eecff1a1df4cc6c858
---
 packages/PackageInstaller/Android.bp                 |  2 +-
 .../com/android/packageinstaller/InstallStart.java   |  3 ++-
 .../packageinstaller/v2/model/InstallRepository.kt   |  8 +++++++-
 .../android/packageinstaller/v2/ui/InstallLaunch.kt  | 10 +++++++++-
 .../v2/ui/fragments/InstallationFragment.java        |  6 ++++++
 .../v2/viewmodel/InstallViewModel.kt                 | 12 +++++++++++-
 6 files changed, 36 insertions(+), 5 deletions(-)

diff --git a/packages/PackageInstaller/Android.bp b/packages/PackageInstaller/Android.bp
index e78dda65bf5b..8027634b7239 100644
--- a/packages/PackageInstaller/Android.bp
+++ b/packages/PackageInstaller/Android.bp
@@ -59,7 +59,7 @@ java_defaults {
         "com.google.android.material_material",
         "kotlin-parcelize-runtime",
     ],
-
+    min_sdk_version: "36",
     lint: {
         error_checks: ["Recycle"],
     },
diff --git a/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java b/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java
index 2c5f965cd16f..129ac441b0c7 100644
--- a/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java
+++ b/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java
@@ -310,7 +310,8 @@ public class InstallStart extends Activity {
     private boolean canPackageQuery(int callingUid, Uri packageUri) {
         ProviderInfo info;
         try {
-            if (Flags.uidBasedProviderLookup()) {
+            if (Flags.uidBasedProviderLookup()
+                    && Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.BAKLAVA_1) {
                 info = mPackageManager.resolveContentProviderForUid(packageUri.getAuthority(),
                     PackageManager.ComponentInfoFlags.of(0), callingUid);
             } else {
diff --git a/packages/PackageInstaller/src/com/android/packageinstaller/v2/model/InstallRepository.kt b/packages/PackageInstaller/src/com/android/packageinstaller/v2/model/InstallRepository.kt
index aee68354fb6b..49e24271482b 100644
--- a/packages/PackageInstaller/src/com/android/packageinstaller/v2/model/InstallRepository.kt
+++ b/packages/PackageInstaller/src/com/android/packageinstaller/v2/model/InstallRepository.kt
@@ -39,19 +39,21 @@ import android.content.pm.PackageInstaller.SessionParams
 import android.content.pm.PackageManager
 import android.content.pm.PackageManager.COMPONENT_ENABLED_STATE_DISABLED
 import android.net.Uri
+import android.os.Build
 import android.os.ParcelFileDescriptor
 import android.os.Process
 import android.os.UserManager
 import android.text.TextUtils
 import android.util.EventLog
 import android.util.Log
+import androidx.annotation.RequiresApi
 import androidx.lifecycle.LiveData
 import androidx.lifecycle.MutableLiveData
 import com.android.packageinstaller.common.EventResultPersister
 import com.android.packageinstaller.common.EventResultPersister.OutOfIdsException
 import com.android.packageinstaller.common.InstallEventReceiver
-import com.android.packageinstaller.stats.StatsdLogger
 import com.android.packageinstaller.stats.PiaStagesLatencyTracker
+import com.android.packageinstaller.stats.StatsdLogger
 import com.android.packageinstaller.v2.model.InstallAborted.Companion.ABORT_REASON_DONE
 import com.android.packageinstaller.v2.model.InstallAborted.Companion.ABORT_REASON_INTERNAL_ERROR
 import com.android.packageinstaller.v2.model.InstallAborted.Companion.ABORT_REASON_POLICY
@@ -867,6 +869,7 @@ class InstallRepository(private val context: Context) : EventResultPersister.Eve
         return true
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     fun requestVerificationConfirmation(): InstallStage {
         var confirmationSnippet: InstallStage = generateConfirmationSnippet()
 
@@ -904,6 +907,7 @@ class InstallRepository(private val context: Context) : EventResultPersister.Eve
         )
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     fun setNegativeVerificationUserResponse(): InstallStage {
         if (PackageInstaller.ACTION_CONFIRM_DEVELOPER_VERIFICATION != intent.action) {
             Log.e(LOG_TAG, "Cannot set verification response for this request: $intent")
@@ -923,6 +927,7 @@ class InstallRepository(private val context: Context) : EventResultPersister.Eve
         )
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     fun setPositiveVerificationUserResponse(): InstallStage {
         if (PackageInstaller.ACTION_CONFIRM_DEVELOPER_VERIFICATION != intent.action) {
             Log.e(LOG_TAG, "Cannot set verification response for this request: $intent")
@@ -946,6 +951,7 @@ class InstallRepository(private val context: Context) : EventResultPersister.Eve
         }
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     fun setRetryVerificationUserResponse(): InstallStage {
         if (PackageInstaller.ACTION_CONFIRM_DEVELOPER_VERIFICATION != intent.action) {
             Log.e(LOG_TAG, "Cannot set verification response for this request: $intent")
diff --git a/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/InstallLaunch.kt b/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/InstallLaunch.kt
index 9040121bbabd..1d7d0f04e3a9 100644
--- a/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/InstallLaunch.kt
+++ b/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/InstallLaunch.kt
@@ -20,6 +20,7 @@ import android.app.AppOpsManager
 import android.content.ActivityNotFoundException
 import android.content.Intent
 import android.net.Uri
+import android.os.Build
 import android.os.Bundle
 import android.os.Handler
 import android.os.Looper
@@ -31,6 +32,7 @@ import android.view.Window
 import android.view.WindowManager
 import androidx.activity.result.ActivityResultLauncher
 import androidx.activity.result.contract.ActivityResultContracts
+import androidx.annotation.RequiresApi
 import androidx.fragment.app.DialogFragment
 import androidx.fragment.app.FragmentActivity
 import androidx.fragment.app.FragmentManager
@@ -49,6 +51,7 @@ import com.android.packageinstaller.v2.ui.fragments.InstallationFragment
 import com.android.packageinstaller.v2.viewmodel.InstallViewModel
 import com.android.packageinstaller.v2.viewmodel.InstallViewModelFactory
 
+@RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA)
 class InstallLaunch : FragmentActivity(), InstallActionListener {
 
     companion object {
@@ -73,7 +76,7 @@ class InstallLaunch : FragmentActivity(), InstallActionListener {
     private var appOpsManager: AppOpsManager? = null
     private lateinit var unknownAppsIntentLauncher: ActivityResultLauncher<Intent>
 
-
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         window.addSystemFlags(WindowManager.LayoutParams.SYSTEM_FLAG_HIDE_NON_SYSTEM_OVERLAY_WINDOWS)
@@ -123,6 +126,7 @@ class InstallLaunch : FragmentActivity(), InstallActionListener {
     /**
      * Main controller of the UI. This method shows relevant dialogs based on the install stage
      */
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     private fun onInstallStageChange(installStage: InstallStage) {
         when (installStage.stageCode) {
             InstallStage.STAGE_STAGING -> {
@@ -249,6 +253,7 @@ class InstallLaunch : FragmentActivity(), InstallActionListener {
         }
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     private fun showInstallationDialog() {
         val fragment = getInstallationFragment() ?: InstallationFragment()
         fragment.updateUI()
@@ -299,6 +304,7 @@ class InstallLaunch : FragmentActivity(), InstallActionListener {
         }
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     override fun onPositiveResponse(reasonCode: Int) {
         if (localLogv) {
             Log.d(LOG_TAG, "Positive button clicked. ReasonCode: $reasonCode")
@@ -313,6 +319,7 @@ class InstallLaunch : FragmentActivity(), InstallActionListener {
         }
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     override fun onNegativeResponse(stageCode: Int) {
         if (localLogv) {
             Log.d(LOG_TAG, "Negative button clicked. StageCode: $stageCode")
@@ -344,6 +351,7 @@ class InstallLaunch : FragmentActivity(), InstallActionListener {
         setResult(resultCode, data, true)
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     override fun onRetryResponse() {
         installViewModel!!.onRetryVerificationUserResponse()
     }
diff --git a/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/fragments/InstallationFragment.java b/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/fragments/InstallationFragment.java
index e32267a20d37..9cd285258520 100644
--- a/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/fragments/InstallationFragment.java
+++ b/packages/PackageInstaller/src/com/android/packageinstaller/v2/ui/fragments/InstallationFragment.java
@@ -28,6 +28,7 @@ import android.content.Intent;
 import android.content.pm.PackageInstaller;
 import android.content.pm.ResolveInfo;
 import android.graphics.Typeface;
+import android.os.Build;
 import android.os.Bundle;
 import android.text.Html;
 import android.text.method.LinkMovementMethod;
@@ -41,6 +42,7 @@ import android.widget.TextView;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import androidx.annotation.RequiresApi;
 import androidx.fragment.app.DialogFragment;
 import androidx.lifecycle.ViewModelProvider;
 
@@ -140,6 +142,7 @@ public class InstallationFragment extends DialogFragment {
     }
 
     @Override
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     public void onStart() {
         super.onStart();
         updateUI();
@@ -169,6 +172,7 @@ public class InstallationFragment extends DialogFragment {
     /**
      * Update the UI based on the current install stage
      */
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     public void updateUI() {
         if (!isAdded()) {
             return;
@@ -492,6 +496,7 @@ public class InstallationFragment extends DialogFragment {
         }
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     private void updateUserActionRequiredUI(Dialog dialog, InstallUserActionRequired installStage) {
         switch (installStage.getActionReason()) {
             case InstallUserActionRequired.USER_ACTION_REASON_INSTALL_CONFIRMATION -> {
@@ -742,6 +747,7 @@ public class InstallationFragment extends DialogFragment {
         }
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     private void updateVerificationConfirmationUI(Dialog dialog,
             InstallUserActionRequired installStage) {
         mAppSnippet.setVisibility(View.VISIBLE);
diff --git a/packages/PackageInstaller/src/com/android/packageinstaller/v2/viewmodel/InstallViewModel.kt b/packages/PackageInstaller/src/com/android/packageinstaller/v2/viewmodel/InstallViewModel.kt
index 7c0194040801..0995f32b624d 100644
--- a/packages/PackageInstaller/src/com/android/packageinstaller/v2/viewmodel/InstallViewModel.kt
+++ b/packages/PackageInstaller/src/com/android/packageinstaller/v2/viewmodel/InstallViewModel.kt
@@ -18,6 +18,8 @@ package com.android.packageinstaller.v2.viewmodel
 
 import android.app.Application
 import android.content.Intent
+import android.os.Build
+import androidx.annotation.RequiresApi
 import androidx.lifecycle.AndroidViewModel
 import androidx.lifecycle.LiveData
 import androidx.lifecycle.MediatorLiveData
@@ -69,7 +71,11 @@ class InstallViewModel(application: Application, val repository: InstallReposito
         ) { installStage: InstallStage ->
             when (installStage.stageCode) {
                 InstallStage.STAGE_READY -> checkIfAllowedAndInitiateInstall()
-                InstallStage.STAGE_VERIFICATION_CONFIRMATION_REQUIRED -> requestVerification()
+                InstallStage.STAGE_VERIFICATION_CONFIRMATION_REQUIRED -> {
+                    if (Build.VERSION.SDK_INT_FULL >= Build.VERSION_CODES_FULL.BAKLAVA_1) {
+                        requestVerification()
+                    }
+                }
                 else -> updateInstallStage(installStage)
             }
         }
@@ -106,22 +112,26 @@ class InstallViewModel(application: Application, val repository: InstallReposito
         }
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     private fun requestVerification() {
         val stage = repository.requestVerificationConfirmation()
         updateInstallStage(stage)
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     fun onNegativeVerificationUserResponse() {
         val stage = repository.setNegativeVerificationUserResponse()
         updateInstallStage(stage)
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     fun onPositiveVerificationUserResponse() {
         val stage =
             repository.setPositiveVerificationUserResponse()
         updateInstallStage(stage)
     }
 
+    @RequiresApi(Build.VERSION_CODES_FULL.BAKLAVA_1)
     fun onRetryVerificationUserResponse() {
         val stage =
             repository.setRetryVerificationUserResponse()
-- 
2.55.0

